src/hotspot/share/gc/g1/g1FullCollector.cpp

```diff
@@ -348,20 +348,44 @@ bool G1FullCollector::phase2b_forward_oops() {
   return task.has_free_compaction_targets();
 }
 
-void G1FullCollector::phase2c_prepare_serial_compaction() {
-  GCTraceTime(Debug, gc, phases) debug("Phase 2: Prepare serial compaction", scope()->timer());
-  // At this point we know that after parallel compaction there will be no
-  // completely free regions. That means that the last region of
-  // all compaction queues still have data in them. We try to compact
-  // these regions in serial to avoid a premature OOM when the mutator wants
-  // to allocate the first eden region after gc.
+void G1FullCollector::add_regions_for_serial_compaction() {
+  // At this point, we know that after parallel compaction there will be regions that
+  // are partially compacted into. Thus, the last compaction region of all
+  // compaction queues still have space in them. We try to re-compact these regions
+  // in serial to avoid a premature OOM when the mutator wants to allocate the first
+  // eden region after gc.
+
+  // For maximum compaction, we need to re-prepare all objects above the lowest
+  // region among the current regions for all thread compaction points. It may
+  // happen that due to the uneven distribution of objects to parallel threads, holes
+  // have been created as threads compact to different target regions between the
+  // lowest and the highest region in the tails of the compaction points.
+  uint lowest_current_hr = (uint)-1;
   for (uint i = 0; i < workers(); i++) {
     G1FullGCCompactionPoint* cp = compaction_point(i);
     if (cp->has_regions()) {
-      serial_compaction_point()->add(cp->remove_last());
+      lowest_current_hr = MIN2(lowest_current_hr, cp->current_region()->hrm_index());
     }
   }
 
+  G1FullGCCompactionPoint* serial_cp = serial_compaction_point();
+  for (uint i = 0; i < workers(); i++) {
+    G1FullGCCompactionPoint* cp = compaction_point(i);
+    if (cp->has_regions()) {
+      cp->move_regions_with_higher_hrm_index(serial_cp, lowest_current_hr);
+    }
+  }
+
+  // We use regions as compaction targets in the order they appear in the compaction
+  // point. To get maximum compaction and reduce fragmentation, we sort the regions
+  // by hrm_index so that we compact objects to one end of the heap.
+  serial_cp->sort_regions();
```

Comment:
I can add a comment if its unclear, however, redoing the filtering as above can be extremely unnecessary, especially where heap has very many pinned regions as in the test-case `vmTestbase/vm/gc/compact/Compact_InternedStrings_NonbranchyTree/TestDescription.java`. You have a few regions for serial compaction, and the sorting is sufficient.  So I would rather keep the current implementation.

Reply:
An alternative, similar to what Albert suggests, would be to change `move_regions_with_higher_hrm_index(...)` to `remove_above(lowest_hr, bitmap)` and when removing the regions from the compaction point store what regions were in use in a bitmap. Then you could use that bitmap to add the regions in order to the serial compaction point.

Reply:
Fwiw, I tend to prefer the `sort` call as it is least amount of code that achieves the goal well enough. All these other suggestions seem to be optimizations for little gain that need lots of extra code/helper data structures for something not (time) critical.

(I may not completely understand @albertnetymk 's suggestion: without either sorting the regions after all of them have been filtered from the worker's compaction point which involves the same sorting, or doing the sorted insertion of the regions for each worker thread's cp regions which is unlikely to be faster than the overall (quick)`sort` at the end, the suggestion does not seem to work).

Reply:
I also prefer the sort. Common case, the number of regions in serial == ~no of threads.

Reply:
> Fwiw, I tend to prefer the sort call as it is least amount of code that achieves the goal well enough. All these other suggestions seem to be optimizations for little gain that need lots of extra code/helper data structures for something not (time) critical.

I think what I suggested requires less code than the sorting approach.

> Common case, the number of regions in serial == ~no of threads

True for now, but when humongous regions can be moved, this number can go up.

Reply:
> I think what I suggested requires less code than the sorting approach.
Unfortunately, that approach is only valid if I add the bitmap as suggest by @kstefanj, otherwise you may iterate over all regions yet the serial compaction has just a few regions as for the test-case I highlighted above.

Reply:
> > Fwiw, I tend to prefer the sort call as it is least amount of code that achieves the goal well enough. All these other suggestions seem to be optimizations for little gain that need lots of extra code/helper data structures for something not (time) critical.
> 
> I think what I suggested requires less code than the sorting approach.
> 

I probably did not understand the pseudo code enough; thinking about it some more, it might work;

The first step is basically dropping all regions >= `lowest_current_hr`; the second about rebuilding the serial cp regions from scratch given the region list (and using some of the existing conditions on whether to add the region to the final serial cp).

The following comes to my mind about this:

* you should not linearly iterate over the discontiguous range of regions using region indexes; there is the public `region_at_or_null` in CollectedHeap to ignore unavailable regions but it's kind of a "don't use" method; code at this level should preferably be oblivious about the regions being contiguous in the heap and holes in the region map.
The only code that uses it this way (`G1RebuildFreeListTask`) is internal to the `HeapRegionManager` (others are to simplify some code or asserts, so maybe that `region_at_or_null` method should be made internal).

* `G1CollectedHeap::heap_region_iterate()` that can be used instead does not guarantee an order of iteration of regions. The current implementation does give you regions in address order though. Not really happy on starting to rely on order of iteration in `heap_region_iterate()` - I am not aware of any other closure doing that (apart of that internal `G1RebuildFreeListTask` which is essentially part of `HeapRegionManager`)

Not sure how I feel about breaking these fairly basic assumptions in how G1 code should think about the heap (particularly the first one).

Reply:
Actually, the existing full gc code probably already breaks the second assumption already when handing out regions round-robin, i.e. assuming that we get regions in increasing address order (but the code would still work without that I guess).

Reply:
I don't have a strong opinion on how to solve this, but I agree with Albert. I think adding a bitmap to keep track of what regions to put in the serial cp and removing them from the others would be less code. But I might be missing some case. I don't see this as an optimization I see it as a more straight forward solution. But we all have different tastes on how to solve things. So I'm good if you want to keep the sort.

Reply:
With the current assumption of ordered region iteration, then Bitmap based solution seems the best to me. It also solves later issues of sorting when adding humongous objects for compaction. We just use the same bitmap to keep track of regions that are available for humongous compaction.

Reply:
After discussions with @albertnetymk, we settled for the latest solution that iterates the `_region_attr_table`.